src/utils/hot_restart.jl

```diff
@@ -0,0 +1,32 @@
+function is_hot_restart_available()
+    return !isinteractive() && !isempty(Base.PROGRAM_FILE) 
+end
+function hot_restart(func::Function; check_interval = 1., env_key = "IS_HOT_RELOADABLE", suppress_warn = false)
+    if !is_hot_restart_available()
+        !suppress_warn && @warn "hot restart is disabled for intereactive sessions"
+        return func()
+    end
+    app_path = abspath(Base.PROGRAM_FILE)
+    if get(ENV, env_key, "false") == "true"
+        @async func()
+        files = parse_includes(app_path)
+        poll_until_changed(files, interval = check_interval)
+        exit(3)
+    else
+        julia_str = joinpath(Sys.BINDIR, Base.julia_exename())
+        new_env = deepcopy(ENV)
+        new_env[env_key] = "true"
+        cmd = Cmd(`$(julia_str) $(app_path)`, env = new_env, ignorestatus = true, windows_hide = true)
+        subp_ref = Ref{Base.Process}()
+        atexit() do
+            if isdefined(subp_ref, 1) && process_running(subp_ref[])
+                kill(subp_ref[])
+            end
+        end
+        while true
+            subp_ref[] = run(pipeline(cmd, stdout = stdout, stderr = stderr), wait = false)
+            wait(subp_ref[])
+            subp_ref[].exitcode != 3 && break
+        end
+    end
```

Comment:
@waralex Am I correct that the implementation of hot reloading vaguely looks like this?
- identify the path to the Julia executable
- copy the current environment
- set a flag `IS_HOT_RELOADABLE` within the new environment
- run the Julia executable using the new environment within a `pipeline`
- kill the previous Julia process

Reply:
@rpkyle Not exactly
if the environment does not have the IS_HOT_RELOADABLE flag:
- identify the path to the Julia executable
- copy the current environment
- set a flag IS_HOT_RELOADABLE within the new environment
- run as subprocess the Julia executable using the new environment within a pipeline
- add killing this subprocess to `atexit` callback
- wait until this subprocess finished. If it finished with exit code 3 - run it again

so process with no IS_HOT_RELOADABLE flag is the "guard" process 

if the environment does have the IS_HOT_RELOADABLE flag: 
- Find recursively all `include` from the executable file and add them to the list of files
- run main function (HTTP.serve etc) asynchronously
- Сontinuously check the time when files from the list were changed, and if any of them changed, finish the process with exit code 3 

so process with IS_HOT_RELOADABLE flag is main work process, which restarts at source files changes

Here is the code used by Flask that gave me the idea for the implementation:
https://github.com/pallets/werkzeug/blob/6868ede8ba4e1f883952154ba68883efac0021a0/src/werkzeug/_reloader.py

Reply:
I've been testing the hot reloading feature, and so far both "soft" (CSS only) and "hard" reloading work perfectly, which is great! The "hard" reload of the backend and restart of the frontend is unfortunately very slow due to Julia's startup time, and the time required to restart the HTTP.jl server. That's understandable, particularly in light of our experience with setting up integration tests and the long timeouts required.

I'm curious if there's a way to avoid restarting Julia, and reload the HTTP.jl server only. In that scenario, the feature would operate more like the way it does in R, where we shutdown `fiery` and refresh the app state while avoiding a server shutdown. In R, we did this because there is no easy way to quit and restart `fiery`, but in Julia a similar approach should make hot reloading much, much faster.

One possibility might be to wrap the handler call that brings up the server within a loop, then send an interrupt to the HTTP.jl server _whenever a reload event occurs_, and recover and launch the server again. If an interrupt occurs outside of a reload event, it can be handled as before, allowing the server to shut down.

Reply:
@rpkyle In a hot restart, you need to recompile all the user code and the Dash code along with it (because the callbacks might have changed, etc.). the safest way to do this is to restart the entire application. What we really need to do is to play with precompilation and optimize the Dash startup in the future

Reply:
@rpkyle 
>One possibility might be to wrap the handler call that brings up the server within a loop, then send an interrupt to the HTTP.jl server whenever a reload event occurs, and recover and launch the server again. If an interrupt occurs outside of a reload event, it can be handled as before, allowing the server to shut down.

I may not have understood the offer, but it doesn't seem to work. Restarting HTTP server won't solve anything because we have to start a DIFFERENT server, with a different application. And this application is defined in the executable file. In any case, we must execute the executable file again.

Reply:
Precompilation isn't really going to resolve this issue completely, since Julia's startup time (before any packages are loaded) is about 70% of the delay at each hot reload cycle. We'll need another solution if we want to substantially reduce the time required to bring the server down and refresh the page.

If there were a way to launch a second Julia process so that it's ready to go, then run the app within it right as the current Julia process is being shut down, that would also improve the time required to cycle (then you'd launch another Julia subprocess, ready to be passed an app in the next reload cycle) -- but I'm not aware of any way to pass code into a Julia subprocess.

Reply:
Let's not spend too much time optimizing this right now - if the feature works, even though it's slow to restart - that's fine for v1.

Reply:
For the moment, I'd lean towards labeling this feature as experimental -- it works as required, but the user experience is a bit different from R or Python and this can be addressed later.

Reply:
@rpkyle, @alexcjohnson I rewrite this in 556f838
Now everything works in one process. Everything is similar to what it was, only the restart is like this - I create a new module with a unique name at the top level and make include the executable file in it. Then everything is also similar to what it was before - when you change any of the source files, the old server stops, a new module is created, and the executable file is re-included in it. This greatly accelerated the restart. 
The probability of pitfalls here is slightly higher than when using brute force-restarting the entire app. But everything seems to be working.

Reply:
The main pitfall that I see is that this approach waits for the entire application to end normally. in Other words, it will not work if there is an infinite loop in the application after calling run_server. But I don't think this is a common case.

Reply:
@waralex The revised approach makes a huge difference! The reload now occurs in about 3 seconds for my sample app, which is a big improvement. 🐎 

Not sure if @alexcjohnson has any feedback regarding this approach, but it now feels much more responsive and comparable to the Dash experience in Python and R.

Reply:
> Now everything works in one process. Everything is similar to what it was, only the restart is like this - I create a new module with a unique name at the top level and make include the executable file in it. 

Neat, I didn't know about the `gensym()` function, that's handy.